Add local disk health checks

klauspost opened the pull request:
## Motivation and Context

The main goal of this PR is to solve the situation where disks stops responding to operations.
This generally causes an FD build-up and eventually will crash the server.

## Description

This adds detection of hung disks, where calls on disk gets stuck.

We add functionality to `xlStorageDiskIDCheck` where it keeps track of the number of concurrent requests on a given disk.

A total number of 100 ongoing operations are allowed. If this limit is reached we will block (but not reject) new requests, but we will monitor the state of the disk.

If no requests have completed or updated without error within a 30 second window, we mark the disk as offline.
Requests that are blocked will be unblocked and return an error. New requests will be rejected until the disk is marked online again. As a safety precaution we never mark a disk offline, where an operation has started within 15 seconds.

Once a disk has been marked offline, a check will run every 5 seconds that will attempt to write and read back a file. As long as this fails the disk will remain offline.

To prevent lots of long-running operations to mark the disk offline we implement a callback feature that allows to update the status as parts of these operations are running.

We add a reader and writer wrapper that will update the status on each successful read/write operation. This should allow fine enough granularity that a slow, but still operational disk will not reach 30 seconds where 50 operations have not progressed.

Note that errors itself is not enough to mark a disk offline. A nil (or io.EOF) error will mark a disk as "good".

Future Improvements:

* Unify the REST server and local xlStorageDiskIDCheck. This would also improve stats significantly.
* Add usage stats, concurrent count, blocked operations, etc.

## How to test this PR?

Testing by inserting an infinite block in `ReadFileStream`, and running `warp get -concurrent=500`.

Stuck ops after running:

```
goroutine profile: total 816
200 @ 0xf1d656 0xf2c5ec 0x31bc6dd 0x3188dbf 0x2e8a62d 0x2f7deb2 0xf4d441
#	0x31bc6dc	github.com/minio/minio/cmd.(*xlStorage).ReadFileStream+0x5c		/minio/cmd/xl-storage.go:1634
#	0x3188dbe	github.com/minio/minio/cmd.(*xlStorageDiskIDCheck).ReadFileStream+0x21e	/minio/cmd/xl-storage-disk-id-check.go:331
#	0x2e8a62c	github.com/minio/minio/cmd.(*streamingBitrotReader).ReadAt+0x16c	/minio/cmd/bitrot-streaming.go:154
#	0x2f7deb1	github.com/minio/minio/cmd.(*parallelReader).Read.func1+0x251		/minio/cmd/erasure-decode.go:165

200 @ 0xf1d656 0xf2c5ec 0x31bc6dd 0x3188dbf 0x315b465 0x302d8b3 0x302d8ae 0x11b286f 0x3025e78 0x11b286f 0x30269e2 0x11b286f 0x3025a0d 0x11b286f 0x3024dae 0x11b286f 0x302402f 0x11b286f 0x2f1edcf 0x11b286f 0x3024343 0x11b286f 0x2e7efd8 0x11b286f 0x11fb18f 0x27959bd 0x11b286f 0x3026b43 0x11b286f 0x121b276 0x11b286f 0x11b5ddb
#	0x31bc6dc	github.com/minio/minio/cmd.(*xlStorage).ReadFileStream+0x5c			/minio/cmd/xl-storage.go:1634
#	0x3188dbe	github.com/minio/minio/cmd.(*xlStorageDiskIDCheck).ReadFileStream+0x21e		/minio/cmd/xl-storage-disk-id-check.go:331
#	0x315b464	github.com/minio/minio/cmd.(*storageRESTServer).ReadFileStreamHandler+0x284	/minio/cmd/storage-rest-server.go:576
#	0x302d8b2	net/http.HandlerFunc.ServeHTTP+0x52						c:/go/src/net/http/server.go:2047 
...
```

There are 4 disks per server. 200 (4x50) are local blocking. 200 are remote calls blocking.

87 GetObject calls are blocked (not shown).

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

klauspost committed:
Add local disk health checks

The main goal of this PR is to solve the situation where disks stops responding to operations.
This generally causes an FD build-up and eventually will crash the server.

This adds detection of hung disks, where calls on disk gets stuck.

We add functionality to `xlStorageDiskIDCheck` where it keeps track of the number of concurrent requests on a given disk.

A total number of 100 operations are allowed. If this limit is reached we will block (but not reject) new requests, but we will monitor the state of the disk.

If no requests have completed or updated within a 15 second window, we mark the disk as offline.
Requests that are blocked will be unblocked and return an error. New requests will be rejected until the disk is marked online again.

Once a disk has been marked offline, a check will run every 5 seconds that will attempt to write and read back a file. As long as this fails the disk will remain offline.

To prevent lots of long-running requests to mark the disk offline we implement a callback feature that allows to update the status as parts of these operations are running.

We add a reader and writer wrapper that will update the status on each successful read/write operation. This should allow fine enough granularity that a slow, but still operational disk will not reach 15 seconds where 100 operations have not progressed.

Note that errors itself is not enough to mark a disk offline. A nil (or io.EOF) error will mark a disk as "good".

Future Improvements:

* Unify the REST server and local xlStorageDiskIDCheck. This would also improve stats significantly.
* Allow reads/writes to be aborted by the context.
* Add usage stats, concurrent count, blocked operations, etc.

klauspost committed:
Add updateStorageMetrics. Do not include blocked time.

klauspost committed:
Add health online check to parent IsOnline().

krisis commented on cmd/xl-storage-disk-id-check.go:
```diff
@@ -579,3 +514,273 @@ func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...st
+	if len(p.health.tokens) < cap(p.health.tokens) {
+		return nil
+	}
+
```
At this point we have all the tokens available for grabs, i.e no ongoing i/o on this disk.
Shouldn't we check the disk's health when `len(p.health.tokens) < cap(p.health.tokens)`. i.e when there is at least one i/o operation in progress?

klauspost replied:
Ah, you are right! It should be `len(p.health.tokens) > 0` @krisis

krisis reviewed (commented):
Great work! I just have one question regarding a condition in checkHealth that I'd like clarified.

klauspost committed:
Use int64 for 32 bit platforms. Check length.

harshavardhana commented on cmd/xl-storage-disk-id-check.go:
```diff
@@ -579,3 +514,276 @@ func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...st
+	diskHealthOffline
+
+	// diskMaxConcurrent is the maximum number of running concurrent operations.
+	diskMaxConcurrent = 100
```
```suggestion
	diskMaxConcurrent = 100
```

we have to make this configurable for testing purposes to arrive at valid values.

harshavardhana commented on cmd/xl-storage-disk-id-check.go:
```diff
@@ -579,3 +514,276 @@ func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...st
+	t := time.NewTicker(5 * time.Second)
+	defer t.Stop()
+	for range t.C {
+		fn := uuid.New().String()
```
mustGetUUID would be better - also we can use same uuid perhaps outside of for loop

harshavardhana reviewed (changes requested):
there are a few things to test for me tomorrow

- replacing disks etc
- also going to put this to a setup for performance tests

klauspost committed:
Stop replacing disk on every monitor cycle.

klauspost committed:
Make concurrent disk setting configurable via `_MINIO_DISK_MAX_CONCURRENT`.

harshavardhana commented on cmd/erasure-sets.go:
```diff
@@ -241,6 +241,7 @@ func (s *erasureSets) connectDisks() {
 			}
 
 			s.erasureDisksMu.Lock()
+			var prevDiskID string
```
This change I don't think is needed it looks like something is wrong in this code which would cause the call to come here @klauspost - you are fixing something wrong AFAICS here.

harshavardhana commented on cmd/xl-storage-disk-id-check.go:
```diff
@@ -579,3 +508,314 @@ func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...st
+					err := *errp
+					ers = fmt.Sprint(err)
+				}
+				fmt.Println(time.Now().Format(time.RFC3339), "op", s, "took", time.Since(t), "result:", ers, "disk:", p.storage.String(), "path:", strings.Join(paths, "/"))
```
is this needed? @klauspost we need to use `console.LogIf()` if needed, sure this is in `if false` perhaps we can use something more standardized than `fmt.Println()`

harshavardhana commented on cmd/xl-storage-disk-id-check.go:
```diff
@@ -579,3 +508,314 @@ func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...st
+var diskMaxConcurrent = 50
+
+func init() {
+	if s, ok := os.LookupEnv("_MINIO_DISK_MAX_CONCURRENT"); ok && s != "" {
```
can we move this to let's say a new sub-system called `drive` and 

```
mc admin config set alias/ drive max-io=100
```

my plan is also to move the `disable-odirect` from API here once we bring it in here..

klauspost replied:
I don't really want this to be user tweakable.

harshavardhana replied:
why not @klauspost we should have such provisions IMO since we do not know which ones work well for NVMe v/s HDDs yet..

harshavardhana replied:
Actually this is not needed, looks like 50 per drive is enough.

harshavardhana committed:
de-couple IsOnline() from disk health tracker

The purpose of IsOnline() is to ensure that we
reconnect the drive only when the "drive" was

- disconnected from network we need to validate
  if the drive is "correct" and is the same drive
  which belongs to this server.

- drive was replaced we have to format it - we
  support hot swapping of the drives.

IsOnline() is not meant for taking the drive offline
when it is hung, it is not useful we can let the
drive be online instead "return" errors for relevant
calls.

harshavardhana reviewed (approved):
LGTM, tested some basic performance tests on 6 node setup all checkout okay.

harshavardhana committed:
return errFaultyDisk for DiskInfo() call

klauspost committed:
Merge branch 'master' into add-disk-health-checks
